### PR TITLE
mesa: Save PCI ID headers, separate mesa_amber

### PR DESCRIPTION
**Pull request by supechicken**
### Changes
- Move `mesa_amber` to a separate file for tidy
- Remove drivers for dGPU (as they are not supported by Chrome OS)
- Save PCI IDs supported by each mesa driver (for setting `MESA_LOADER_DRIVER_OVERRIDE` in sommelier)
- Cleanup

Needs binary, tested on `x86_64`

### Run the following to get this pull request's changes locally for testing.
```
CREW_TESTING_REPO=https://github.com/supechicken/chromebrew.git CREW_TESTING_BRANCH=mesa_cleanup CREW_TESTING=1 crew update
```

### CODE REVIEW

**Comment by satmandu** [comment]
Do you not need the gbm symlinks for arm for mesa_amber?

**Comment by satmandu** [comment]
And how does this handle a device getting updated to a newer kernel where you want mesa instead of mesa_amber?

File: packages/mesa.rb [hunk 1]
```diff
@@ -1,373 +1,125 @@
...
+# We use mesa amber (derived from the 21.3 series) for intel hardware with older kernels
+# and current mesa versions for newer kernels.
+if !CREW_IN_CONTAINER && CREW_IS_INTEL && ( Gem::Version.new(CREW_KERNEL_VERSION) < Gem::Version.new('4.16') )
+  Package.load_package( File.join(CREW_PACKAGES_PATH, 'mesa_amber.rb') )
+  class Mesa < Package::Mesa_amber
+    def self.preflight
+      warn 'Installing mesa_amber as kernel is older than 4.16'.lightcyan
+    end
+  end
+else
```
Comment (uberhacker) [inline packages/mesa.rb:12]: This looks very hackish.  Would it make sense to add this logic to preflight instead and suggest to install the other package with `crew install mesa|mesa_amber`?

Comment (supechicken) [reply packages/mesa.rb:12]: > Also, what do we do about all the packages that have `depends_on 'mesa'`? This logic would have to be applied there also, correct?

Yep

> This looks very hackish. Would it make sense to add this logic to preflight instead and suggest to install the other package with `crew install mesa|mesa_amber`?

Another way could be turning `mesa.rb` into a fake package and break down into `mesa_latest.rb` and `mesa_amber.rb` (it seems a better approach :) )

**Comment by supechicken** [comment]
> Do you not need the gbm symlinks for arm for mesa_amber?

Since they are still supported by the latest mesa (as I know in mainline no drivers required by arm chromebooks were dropped`)

**Comment by satmandu** [comment]
> > Do you not need the gbm symlinks for arm for mesa_amber?
> 
> Since the GPU on almost all arm Chromebooks are still supported by the latest mesa, we don't need to use `mesa-amber` on these platforms (as I know in mainline no drivers required by arm Chromebooks were dropped)

You still need to create the symlinks. Sommelier otherwise won't find the GBM files.

File: packages/mesa.rb (hunk 1 above)
Comment (satmandu) [reply packages/mesa.rb:12]: With Mesa 23.0.x coming out this week, would it be a good time to make this sort of change now?

File: packages/mesa_amber.rb [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    system 'patch', '-Np1', '-F', '10', '-i', '16289.diff'
+    system 'patch', '-Np1', '-i', '17514.diff'
+    system 'patch', '-Np1', '-i', '17518.patch'
+  end
+
+  def self.build
+    system <<~BUILD
+      mold -run meson setup #{CREW_MESON_OPTIONS} \
+        -Db_asneeded=false \
+        -Damber=true \
```
Comment (satmandu) [inline packages/mesa_amber.rb:209]: If I recall correctly setting `-Damber=true` kept certain drivers from being built, which kept sommelier from working. Do you see that at all?

**Comment by satmandu** [comment]
This is my current working mesa 23.x package:
```
require 'package'

class Mesa < Package
  description 'Open-source implementation of the OpenGL specification'
  homepage 'https://www.mesa3d.org'
  # We use mesa amber (derived from the 21.3 series) for older kernels
  # and current mesa versions for newer kernels.
  if Gem::Version.new(CREW_KERNEL_VERSION.to_s) < Gem::Version.new('5.10')
    # Built off of the mesa amber branch
    git_hashtag '14d7da24640dadfb55769cfc93e13ab653996fed'
    @_ver = git_hashtag[0, 7]
    version "amber-#{@_ver}"
  else
    @_ver = '23.0.0-rc4'
    version @_ver
    git_hashtag "mesa-#{@_ver}"
  end
  license 'MIT'
  compatibility 'all'
  source_url 'https://gitlab.freedesktop.org/mesa/mesa.git'

  if Gem::Version.new(CREW_KERNEL_VERSION.to_s) < Gem::Version.new('5.10')
    binary_url({
      aarch64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/mesa/amber-acfef00_armv7l/mesa-amber-acfef00-chromeos-armv7l.tar.zst',
       armv7l: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/mesa/amber-acfef00_armv7l/mesa-amber-acfef00-chromeos-armv7l.tar.zst',
         i686: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/mesa/amber-acfef00_i686/mesa-amber-acfef00-chromeos-i686.tar.zst',
       x86_64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/mesa/amber-acfef00_x86_64/mesa-amber-acfef00-chromeos-x86_64.tar.zst'
    })
    binary_sha256({
      aarch64: 'e9ef6ce22956bdae6f75a44e953cc189f08a3c7e7cf9235a24dbe77f94dc39dd',
       armv7l: 'e9ef6ce22956bdae6f75a44e953cc189f08a3c7e7cf9235a24dbe77f94dc39dd',
         i686: '596ce6a08d3b63ba990058acdf876e4b815882e3d0e0c61bfcd079fdfeb6fb9f',
       x86_64: '9fe6dfd910312956da149dd5cd95f2fd58f7ca4bd8a4dbc646ec11ac3864a683'
    })
  else
    binary_url({
      aarch64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/mesa/23.0.0-rc4_armv7l/mesa-23.0.0-rc4-chromeos-armv7l.tar.zst',
       armv7l: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/mesa/23.0.0-rc4_armv7l/mesa-23.0.0-rc4-chromeos-armv7l.tar.zst',
       x86_64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/mesa/23.0.0-rc4_x86_64/mesa-23.0.0-rc4-chromeos-x86_64.tar.zst'
    })
    binary_sha256({
      aarch64: 'b85c9f423e511465f2276f9419924c7471e2f7b01aae8430f64580fa78377f62',
       armv7l: 'b85c9f423e511465f2276f9419924c7471e2f7b01aae8430f64580fa78377f62',
       x86_64: '8c7c5ab04bfb68ed41496e7cc31094bd9dbd994f8a4fcd50dc4ae0aaa3f3d6af'
    })
  end

  depends_on 'elfutils' # R
  depends_on 'eudev' # R
  depends_on 'expat' # R
  depends_on 'gcc' # R
  depends_on 'glibc' # R
  depends_on 'glslang' => :build
  depends_on 'libdrm' # R
  depends_on 'libglvnd' # R
  depends_on 'libomxil_bellagio' => :build
  depends_on 'libunwind'
  depends_on 'libva' => :build # Enable only during build to avoid circular dep.
  depends_on 'libvdpau' => :build
  depends_on 'libx11' # R
  depends_on 'libxcb' # R
  depends_on 'libxdamage' => :build
  depends_on 'libxdmcp' => :build
  depends_on 'libxext' # R
  depends_on 'libxfixes' # R
  depends_on 'libxrandr' # R
  depends_on 'libxshmfence' # R
  depends_on 'libxv' => :build
  depends_on 'libxvmc' # R
  depends_on 'libxv' # R
  depends_on 'libxxf86vm' # R
  depends_on 'llvm' => :build
  depends_on 'lm_sensors' # R
  depends_on 'py3_mako'
  depends_on 'valgrind' => :build
  depends_on 'vulkan_headers' => :build
  depends_on 'vulkan_icd_loader' => :build
  depends_on 'vulkan_icd_loader' # R
  depends_on 'wayland_protocols' => :build
  depends_on 'wayland' # R
  depends_on 'zlibpkg' # R
  depends_on 'zstd' # R

  if Gem::Version.new(CREW_KERNEL_VERSION.to_s) < Gem::Version.new('5.10')
    def self.patch
      puts 'Downloading mesa amber patches...'.yellow
      case ARCH
      when 'aarch64', 'armv7l'
        # See https://gitlab.freedesktop.org/mesa/mesa/-/issues/5067
        @freedrenopatch = <<~FREEDRENOPATCHEOF
                  --- a/src/gallium/drivers/freedreno/freedreno_util.h   2021-08-05 14:40:22.000000000 +0000
                  +++ b/src/gallium/drivers/freedreno/freedreno_util.h   2021-08-05 19:52:53.115410668 +0000
                  @@ -44,6 +44,15 @@
                   #include "adreno_pm4.xml.h"
                   #include "disasm.h"
          #{'         '}
                  +#include <unistd.h>
                  +#include <sys/syscall.h>
                  +
                  +#ifndef SYS_gettid
                  +#error "SYS_gettid unavailable on this system"
                  +#endif
                  +
                  +#define gettid() ((pid_t)syscall(SYS_gettid))
                  +
                   #ifdef __cplusplus
                   extern "C" {
                   #endif
        FREEDRENOPATCHEOF
        File.write('freedreno.patch', @freedrenopatch)
        # See https://gitlab.freedesktop.org/mesa/mesa/-/issues/3505
        @tegrapatch = <<~TEGRAPATCHEOF
                  diff --git a/src/gallium/drivers/nouveau/nvc0/nvc0_state_validate.c b/src/gallium/drivers/nouveau/nvc0/nvc0_state_validate.c
                  index 48d81f197db..f9b7bd57b27 100644
                  --- a/src/gallium/drivers/nouveau/nvc0/nvc0_state_validate.c
                  +++ b/src/gallium/drivers/nouveau/nvc0/nvc0_state_validate.c
                  @@ -255,6 +255,10 @@ nvc0_validate_fb(struct nvc0_context *nvc0)
          #{'         '}
                            nvc0_resource_fence(res, NOUVEAU_BO_WR);
          #{'         '}
                  +         // hack to make opengl at least halfway working on a tegra k1
                  +         // see: https://gitlab.freedesktop.org/mesa/mesa/-/issues/3505#note_627006
                  +         fb->zsbuf=NULL;
                  +
                            assert(!fb->zsbuf);
                         }
        TEGRAPATCHEOF
        File.write('tegra.patch', @tegrapatch)

        system 'patch -Np1 -i freedreno.patch'
        system 'patch -Np1 -i tegra.patch'
      end
      # llvm 13/14 patch  See https://gitlab.freedesktop.org/mesa/mesa/-/issues/5455
      # & https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/13273
      downloader 'https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/13273.diff',
                 '76d2dd16532336837bccd4885c40efed0ab5f1de8e8fa114a7835dc269f221ac'
      # mesa: Implement ANGLE_sync_control_rate (used by Chrome browser)
      downloader 'https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/15381.diff',
                 '1391e189f5ad40a711a6f72a7d59aef1b943ec9dc408852f5f562699bf50ba6c'
      # The following patches are all for llvm 15:
      downloader 'https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/15091.diff',
                 'c53387c9fce1f34b6d7c0272ebef148dda59dea35fd83df2f3f4a0033732ebbd'
      downloader 'https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/15232.diff',
                 'c66b6b03a59ad43a89bc7ab4e04f8c311631d27c3ea6769217c09beef707d6c3'
      downloader 'https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/16129.diff',
                 '88e5d7f6b4e6dd4ac7220cf194aab6e86d748a8cb99a86515eb4c6bdf9b20959'
      downloader 'https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/16289.diff',
                  '56725f4238d8bb60d813db1724e37bf149345ff456c0c2792f0982d237c18cf1'
      downloader 'https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/17514.diff',
                 'b769f0eb2db0b71723f8ad6f20c03a166a54eab74bfd292cf5b9c8ea86d2c73b'
      # Refreshed llvm15 patch from https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/17518.diff
      @mesa_17518_patch = <<~'PATCH_EOF'
        diff -Npaur a/lp_bld_arit.c b/lp_bld_arit.c
        --- a/src/gallium/auxiliary/gallivm/lp_bld_arit.c
        +++ b/src/gallium/auxiliary/gallivm/lp_bld_arit.c
        @@ -391,16 +391,10 @@ lp_build_comp(struct lp_build_context *b
                  return LLVMBuildNot(builder, a, "");
            }

        -   if(LLVMIsConstant(a))
        -      if (type.floating)
        -          return LLVMConstFSub(bld->one, a);
        -      else
        -          return LLVMConstSub(bld->one, a);
        +   if (type.floating)
        +      return LLVMBuildFSub(builder, bld->one, a, "");
            else
        -      if (type.floating)
        -         return LLVMBuildFSub(builder, bld->one, a, "");
        -      else
        -         return LLVMBuildSub(builder, bld->one, a, "");
        +      return LLVMBuildSub(builder, bld->one, a, "");
         }


        @@ -479,16 +473,10 @@ lp_build_add(struct lp_build_context *bl
               }
            }

        -   if(LLVMIsConstant(a) && LLVMIsConstant(b))
        -      if (type.floating)
        -         res = LLVMConstFAdd(a, b);
        -      else
        -         res = LLVMConstAdd(a, b);
        +   if (type.floating)
        +      res = LLVMBuildFAdd(builder, a, b, "");
            else
        -      if (type.floating)
        -         res = LLVMBuildFAdd(builder, a, b, "");
        -      else
        -         res = LLVMBuildAdd(builder, a, b, "");
        +      res = LLVMBuildAdd(builder, a, b, "");

            /* clamp to ceiling of 1.0 */
            if(bld->type.norm && (bld->type.floating || bld->type.fixed))
        @@ -815,16 +803,10 @@ lp_build_sub(struct lp_build_context *bl
               }
            }

        -   if(LLVMIsConstant(a) && LLVMIsConstant(b))
        -      if (type.floating)
        -         res = LLVMConstFSub(a, b);
        -      else
        -         res = LLVMConstSub(a, b);
        +   if (type.floating)
        +      res = LLVMBuildFSub(builder, a, b, "");
            else
        -      if (type.floating)
        -         res = LLVMBuildFSub(builder, a, b, "");
        -      else
        -         res = LLVMBuildSub(builder, a, b, "");
        +      res = LLVMBuildSub(builder, a, b, "");

            if(bld->type.norm && (bld->type.floating || bld->type.fixed))
               res = lp_build_max_simple(bld, res, bld->zero, GALLIVM_NAN_RETURN_OTHER_SECOND_NONNAN);
        @@ -980,29 +962,15 @@ lp_build_mul(struct lp_build_context *bl
            else
               shift = NULL;

        -   if(LLVMIsConstant(a) && LLVMIsConstant(b)) {
        -      if (type.floating)
        -         res = LLVMConstFMul(a, b);
        -      else
        -         res = LLVMConstMul(a, b);
        -      if(shift) {
        -         if(type.sign)
        -            res = LLVMConstAShr(res, shift);
        -         else
        -            res = LLVMConstLShr(res, shift);
        -      }
        -   }
        -   else {
        -      if (type.floating)
        -         res = LLVMBuildFMul(builder, a, b, "");
        +   if (type.floating)
        +       res = LLVMBuildFMul(builder, a, b, "");
        +    else
        +       res = LLVMBuildMul(builder, a, b, "");
        +    if (shift) {
        +       if (type.sign)
        +          res = LLVMBuildAShr(builder, res, shift, "");
               else
        -         res = LLVMBuildMul(builder, a, b, "");
        -      if(shift) {
        -         if(type.sign)
        -            res = LLVMBuildAShr(builder, res, shift, "");
        -         else
        -            res = LLVMBuildLShr(builder, res, shift, "");
        -      }
        +          res = LLVMBuildLShr(builder, res, shift, "");
            }

            return res;
        @@ -1288,15 +1256,6 @@ lp_build_div(struct lp_build_context *bl
            if(a == bld->undef || b == bld->undef)
               return bld->undef;

        -   if(LLVMIsConstant(a) && LLVMIsConstant(b)) {
        -      if (type.floating)
        -         return LLVMConstFDiv(a, b);
        -      else if (type.sign)
        -         return LLVMConstSDiv(a, b);
        -      else
        -         return LLVMConstUDiv(a, b);
        -   }
        -
            /* fast rcp is disabled (just uses div), so makes no sense to try that */
            if(FALSE &&
               ((util_get_cpu_caps()->has_sse && type.width == 32 && type.length == 4) ||
        @@ -2643,7 +2602,7 @@ lp_build_rcp(struct lp_build_context *bl
            assert(type.floating);

            if(LLVMIsConstant(a))
        -      return LLVMConstFDiv(bld->one, a);
        +      return LLVMBuildFDiv(builder, bld->one, a, "");

            /*
             * We don't use RCPPS because:

      PATCH_EOF
      File.write('17518.patch', @mesa_17518_patch)
      system 'patch', '-Np1', '-i', '13273.diff'
      system 'patch', '-Np1', '-i', '15381.diff'
      system 'patch', '-Np1', '-i', '15091.diff'
      system 'patch', '-Np1', '-i', '15232.diff'
      system 'patch', '-Np1', '-i', '16129.diff'
      system 'patch', '-Np1', '-F', '10', '-i', '16289.diff'
      system 'patch', '-Np1', '-i', '17514.diff'
      system 'patch', '-Np1', '-i', '17518.patch'
    end

    def self.build
      case ARCH
      when 'i686'
        @vulkan_drivers_amber = 'intel,swrast'
        @gallium_drivers_amber = 'swrast,svga,virgl,zink'
        @lto_amber_meson_options = CREW_MESON_FNO_LTO_OPTIONS
        @osmesa_amber = 'false'
      when 'aarch64', 'armv7l'
        @vulkan_drivers_amber = 'auto'
        @gallium_drivers_amber = 'auto'
        lto_amber_meson_options = CREW_MESON_OPTIONS
        @osmesa_amber = 'true'
      when 'x86_64'
        @vulkan_drivers_amber = 'auto'
        @gallium_drivers_amber = 'r300,r600,radeonsi,nouveau,virgl,svga,swrast,iris,crocus,zink'
        @lto_amber_meson_options = CREW_MESON_OPTIONS
        @osmesa_amber = 'true'
      end
      system "meson setup #{lto_amber_meson_options} \
      -Damber=true \
      -Db_asneeded=false \
      -Ddri3=enabled \
      -Ddri-drivers=auto \
      -Degl=enabled \
      -Dgallium-drivers=#{@gallium_drivers_amber} \
      -Dgbm=enabled \
      -Dgles2=enabled \
      -Dglvnd=true \
      -Dosmesa=#{@osmesa_amber} \
      -Dshared-glapi=enabled \
      -Dvulkan-drivers=#{@vulkan_drivers_amber} \
       builddir"
      system 'meson configure builddir'
      system 'mold -run samu -C builddir'
    end
  else
    def self.build
      @gallium_drivers = ARCH == 'x86_64' ? 'i915,r300,r600,radeonsi,nouveau,virgl,svga,swrast,iris,crocus,zink' : 'auto'
      @vulkan_drivers = ARCH == 'x86_64' ? 'amd, intel, intel_hasvk, swrast' : 'auto'
      system "meson setup #{CREW_MESON_OPTIONS} \
      -Db_asneeded=false \
      -Ddri3=enabled \
      -Degl=enabled \
      -Dgbm=enabled \
      -Dgles1=disabled \
      -Dgles2=enabled \
      -Dglvnd=true \
      -Dglx=dri \
      -Dllvm=enabled \
      -Dgallium-drivers='#{@gallium_drivers}' \
      -Dvulkan-drivers='#{@vulkan_drivers}' \
      -Dvideo-codecs='vc1dec,h264dec,h264enc,h265dec,h265enc' \
       builddir"
      system 'meson configure builddir'
      system 'mold -run samu -C builddir'
    end
  end

  def self.install
    system "DESTDIR=#{CREW_DEST_DIR} samu -C builddir install"
    # The following are hacks to keep sommelier from complaining.
    Dir.chdir("#{CREW_DEST_LIB_PREFIX}/dri") do
      FileUtils.ln_s '.', 'tls' unless File.exist?('tls')
    end
    FileUtils.mkdir_p "#{CREW_DEST_LIB_PREFIX}/gbm/tls"
    case ARCH
    when 'x86_64', 'i686'
      Dir.chdir("#{CREW_DEST_LIB_PREFIX}/gbm/tls") do
        # For Intel GPUs
        FileUtils.ln_s '../../libgbm.so', 'i915_gbm.so'
        # For AMD GPUs
        FileUtils.ln_s '../../libgbm.so', 'amdgpu_gbm.so'
      end
    when 'armv7l', 'aarch64'
      Dir.chdir("#{CREW_DEST_LIB_PREFIX}/gbm/tls") do
        FileUtils.ln_s '../../libgbm.so', 'pvr_gbm.so'
      end
    end
  end
end
```

**Comment by satmandu** [comment]
@supechicken do you get acceleration with the mesa in https://github.com/chromebrew/chromebrew/pull/7938 ?

**Comment by satmandu** [comment]
Same question for @uberhacker . Do you see acceleration with `glxinfo -B` on either `armv7l` or `aarch64`...

**Comment by uberhacker** [comment]
> Same question for @uberhacker . Do you see acceleration with `glxinfo -B` on either `armv7l` or `aarch64`...

I get the same output on armv7l and aarch64.  See below:
```
$ glxinfo -B
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Mesa/X.org (0xffffffff)
    Device: llvmpipe (LLVM 15.0.7, 128 bits) (0xffffffff)
    Version: 21.3.9
    Accelerated: no
    Video memory: 3936MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 4.5
    Max compat profile version: 4.5
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.2
OpenGL vendor string: Mesa/X.org
OpenGL renderer string: llvmpipe (LLVM 15.0.7, 128 bits)
OpenGL core profile version string: 4.5 (Core Profile) Mesa 21.3.9 (git-14d7da2464)
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 4.5 (Compatibility Profile) Mesa 21.3.9 (git-14d7da2464)
OpenGL shading language version string: 4.50
OpenGL context flags: (none)
OpenGL profile mask: compatibility profile

OpenGL ES profile version string: OpenGL ES 3.2 Mesa 21.3.9 (git-14d7da2464)
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
```

**Comment by satmandu** [comment]
@supechicken looks like we're at least getting opengl working fine on arm, without acceleration. Would there be any other advantage to the changes here for any architecture? (I don't think this PR actually enabled acceleration on ARM...)

**Comment by supechicken** [comment]
> @supechicken looks like we're at least getting opengl working fine on arm, without acceleration. Would there be any other advantage to the changes here for any architecture? (I don't think this PR actually enabled acceleration on ARM...)

No, this PR only targets on separating the mainline mesa and amber mesa

**Pull request closed without merging**